### PR TITLE
ADM 434 Jira board calculate CycleTime

### PR DESCRIPTION
**Pull request by Eason-T**
## Summary
- calculate CycleTime method
- fix WorkDay duplicate requests to holiday API
...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/main/java/heartbeat/controller/report/dto/response/CycleTime.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+@AllArgsConstructor
+public class CycleTime {
+
+	private double totalTime;
+
+	private double averageCircleTimePerCard;
+
+	private double averageCycleTimePerSP;
+
+	private List<CycleTimeForSelectedStepItem> cycleTimeForSelectedStepsList;
```
Comment (wangwang-zhang) [inline backend/src/main/java/heartbeat/controller/report/dto/response/CycleTime.java:22]: cycleTimeForSelectedStepsList to cycleTimeForSelectedSteps?

File: backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			cycleTimeForSelectedStepsList.add(cycleTimeOptionalItem);
+			totalTime += value;
+		}
+		return CycleTimeResult.builder()
+			.cycleTimeForSelectedStepsList(cycleTimeForSelectedStepsList)
+			.totalTime(Double.parseDouble(String.format("%.2f", totalTime)))
+			.build();
+	}
+
+	public CycleTime calculateCycleTime(CardCollection cardCollection,
```
Comment (wangwang-zhang) [inline backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java:96]: move public function to top

**Review by wangwang-zhang — APPROVED** [review]

File: backend/src/main/java/heartbeat/controller/report/dto/response/CycleTime.java (hunk 1 above)
Comment (Eason-T) [reply backend/src/main/java/heartbeat/controller/report/dto/response/CycleTime.java]: done

File: backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java (hunk 2 above)
Comment (Eason-T) [reply backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java]: done

File: backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@RequiredArgsConstructor
+@Component
+public class CycleTimeCalculator {
+
+	public CycleTime calculateCycleTime(CardCollection cardCollection,
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java:21]: `calculateCycleTime` likes the class name, should use `calculate`

File: backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+			List<RequestJiraBoardColumnSetting> boardColumns) {
+		Map<String, String> selectedStepsMap = selectedStepsArrayToMap(boardColumns);
+		Map<String, Double> totalTimeOfEachStepsMap = addAllCardsTimeUpForEachStep(cardCollection, selectedStepsMap);
+		Map<String, Double> aggregatedMap = aggregateResultBySelectedSteps(totalTimeOfEachStepsMap, selectedStepsMap);
+		CycleTimeResult cycleTimeResult = calculateAverageTimeAndTotalTime(aggregatedMap, cardCollection);
+		double cycleTotalTime = cycleTimeResult.getTotalTime();
+		return CycleTime.builder()
+			.totalTime(cycleTotalTime)
+			.averageCycleTimePerSP(
+					Double.parseDouble(String.format("%.2f", cycleTotalTime / cardCollection.getStoryPointSum())))
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java:31]: refactor `String.format("%.2f", c` into a utils method or a static method

File: backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		for (Map.Entry<String, Double> entry : aggregatedMap.entrySet()) {
+			String key = entry.getKey();
+			double value = entry.getValue();
+			if (List.of(CardStepsEnum.ANALYSE, CardStepsEnum.TODO, CardStepsEnum.DONE)
+				.contains(CardStepsEnum.fromValue(key))) {
+				continue;
+			}
+			CycleTimeForSelectedStepItem cycleTimeOptionalItem = CycleTimeForSelectedStepItem.builder()
+				.optionalItemName(key)
+				.averageTimeForSP(Double.parseDouble(String.format("%.2f", value / cardCollection.getStoryPointSum())))
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java:99]: `String.format("%.2f",`

File: backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+				.averageTimeForCards(Double.parseDouble(String.format("%.2f", value / cardCollection.getCardsNumber())))
+				.totalTime(value)
+				.build();
+
+			cycleTimeForSelectedStepsList.add(cycleTimeOptionalItem);
+			totalTime += value;
+		}
+		return CycleTimeResult.builder()
+			.cycleTimeForSelectedStepsList(cycleTimeForSelectedStepsList)
+			.totalTime(Double.parseDouble(String.format("%.2f", totalTime)))
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/calculator/CycleTimeCalculator.java:109]: String.format("%.2f", tota

**Pull request merged**